### ISSUE TITLE
Process JSON array fields

### ISSUE THREAD

**Issue by jesuslinares** [wazuh/wazuh#2013]
Hi team,

If an event has a field with an array of objects, the field is removed. Example:

The following event:

> {"field1": [ {"a": 1, "b": 2}, {"a": 3, "b": 4} ], "field2": "attack"  }

causes this alert:
> {"field2": "attack" }

We should allow nested objects.

On the other hand, the proper datatype must be used in Elaticsearch to work with this kind of fields: https://www.elastic.co/guide/en/elasticsearch/reference/current/nested.html.

**Comment by crolopez** [comment]
Solved in https://github.com/wazuh/wazuh/pull/2016.

**Comment by vikman90** [comment]
This issue needs some discussion: dynamic fields don't support any kind of arrays at this moment.

We are looking forward to supporting full JSON objects, including nested objects and arrays. This is the current behavior:

- Nested objects: The subkey is appended to the parent key, dot-separated.
- Arrays of numbers or strings: The content is gathered into a single string, comma-separated.
- Arrays of objects: Not supported currently.

Mapping the array index into a key may be a solution, but rules based on dynamic fields will need an exact index identifier.

**Comment by branchnetconsulting** [comment]
Could we please revisit establishing some kind of support for parsing JSON records containing arrays of objects?  I know "mapping the array index into a key" is not an ideal solution in that Wazuh rules can't refer to wildcarded field names, and it puts such fields out of reach of simple Kibana aggregations and visualizations, but at least it would parse out all of the content of our JSON log records instead of leaving potentially important elements discarded or buried in the middle of a huge full_log JSON string field.  With Kibana 7.x we now can make queries with a wildcard in the field name, so I believe something like this would be worth doing.  There are important kinds of JSON logs like Office 365 and Kubernetes auditsink, that Wazuh really needs to be able to parse one way or another.  

**Like if Wazuh encounters this:**

{"Emails": [ {"MessageID": "AfMt0C1Xl", "Subject": "Word News", "Path": "\\Inbox"}, {"MessageID": "C1XlBwBQE", "Subject": "Big Sale", "Path": "\\Inbox"} ] }

```
{
  "Emails": [
    {
      "MessageID": "AfMt0C1Xl",
      "Subject": "Word News",
      "Path": "\\Inbox"
    },
    {
      "MessageID": "C1XlBwBQE",
      "Subject": "Big Sale",
      "Path": "\\Inbox"
    }
  ]
}
```

**It could automatically detect the array contains objects and convert it like this:**

{"Emails#1": {"MessageID": "AfMt0C1Xl", "Subject": "Word News", "Path": "\\Inbox"}, "Emails#2": {"MessageID": "C1XlBwBQE", "Subject": "Big Sale", "Path": "\\Inbox"} }

```
{
  "Emails#1": {
    "MessageID": "AfMt0C1Xl",
    "Subject": "Word News",
    "Path": "\\Inbox"
  },
  "Emails#2": {
    "MessageID": "C1XlBwBQE",
    "Subject": "Big Sale",
    "Path": "\\Inbox"
  }
}
```
With the new indexes objects it would then be possible to use Kibana 7.x to query for something like:

**Comment by vikman90** [comment]
@nicutor This issue will be addressed in #11334 as the new engine will fully support JSON, both in data entry and rule handling.

Best regards.
